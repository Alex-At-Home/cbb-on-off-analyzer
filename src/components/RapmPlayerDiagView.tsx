// React imports:
import React, { useState } from 'react';

import _ from "lodash";

// Next imports:
import { NextPage } from 'next';

// Utils
import { RapmInfo, RapmPlayerContext, RapmPreProcDiagnostics, RapmProcessingInputs, RapmUtils } from "../utils/stats/RapmUtils";

type Props = {
  rapmInfo: RapmInfo,
  player: Record<string, any>,
  globalRef: React.RefObject<HTMLDivElement>
};

const RapmPlayerDiagView: React.FunctionComponent<Props> = (({rapmInfo, player, globalRef}) => {
  try {
    const ctx = rapmInfo.ctx;
    const offWeights = rapmInfo.offWeights.valueOf();
    const offInputs = rapmInfo.offInputs;
    const defInputs = rapmInfo.defInputs;

    const gotoGlobalDiags = () => {
      if (globalRef.current) {
        globalRef.current.scrollIntoView({
          behavior: 'smooth',
          block: 'start',
        });
      }
    };

    const rapmOff = (player.rapm?.off_adj_ppp?.value || 0);
    const rapmDef = (player.rapm?.def_adj_ppp?.value || 0);
    const calcPriorDiags = () => {
      if (ctx.unbiasWeight > 0) {
        const totalOffPoss = ctx.teamInfo?.off_poss?.value;
        const teamOffAdj = ((ctx.teamInfo.off_adj_ppp?.value || ctx.avgEfficiency) - ctx.avgEfficiency);
        const teamDefAdj = ((ctx.teamInfo.def_adj_ppp?.value || ctx.avgEfficiency) - ctx.avgEfficiency);

        const col = ctx.playerToCol[player.playerId];
        const offPossPcts = offWeights[ctx.numLineups]; //(use as approx for def also)
        const offPossPctStr = (100.0*(offPossPcts[col]/ctx.unbiasWeight)).toFixed(0);

        const [ sigmaRapmOff, sigmaRapmDef ] = rapmInfo?.noUnbiasWeightsDiags?.map((vec) =>
          _.reduce(vec, (acc, n: number, i: number) => acc + n*offPossPcts[i]/ctx.unbiasWeight) || 0
        ) || [0, 0];
        const offPriorTotalDiff = teamOffAdj - sigmaRapmOff;
        const defPriorTotalDiff = teamDefAdj - sigmaRapmDef;

        const [ offUnbiasRapm, defUnbiasRapm ] = rapmInfo?.noUnbiasWeightsDiags?.map((vec) => vec[col]) || [0, 0];
        const offPrior = rapmOff - offUnbiasRapm;
        const defPrior = rapmDef - defUnbiasRapm;

        return [
          offPrior, defPrior,
          <ul>
            <li>We calculate the total team priors (off=[<b>{offPriorTotalDiff.toFixed(2)}</b>] def=[<b>{defPriorTotalDiff.toFixed(2)}</b>]) from
            the reduction in the team adjusted efficiency due to the regression factor, eg compare <em>observed</em> (off=[<b>{teamOffAdj.toFixed(2)}</b>] def=[<b>{teamDefAdj.toFixed(2)}</b>])
            vs <em>derived from RAPM</em> (off=[<b>{sigmaRapmOff.toFixed(2)}</b>] def=[<b>{sigmaRapmDef.toFixed(2)}</b>]).
            </li>
            <li>Then we calculate a player's contribution to the team prior - currently this is mostly based on their % on floor, [<b>{offPossPctStr}%</b>] (of [<b>{totalOffPoss}</b>] poss);
            eg for the offense approximately [<b>{offPossPctStr}%</b>] * (1/5) * [<b>{offPriorTotalDiff.toFixed(2)}</b>] = [<b>{(0.2*offPossPcts[col]*offPriorTotalDiff/ctx.unbiasWeight).toFixed(2)}</b>],
            plus some smaller lineup adjustments.
            </li>
          </ul>
        ];

      } else {
        return [ 0, 0, <ul><li>Not currently supported</li></ul> ]; //(not currently supported)
      }
    };
    const [ priorOff, priorDef, priorDiagListEl ] = calcPriorDiags() as [ number, number, Element ];
    const totalPrior = priorOff - priorDef;
    const rawRapmOff = rapmOff - priorOff;
    const rawRapmDef = rapmDef - priorDef;
    const totalRawRapm = rawRapmOff - rawRapmDef;

    return <span>
        <b>RAPM diagnostics for [{player.playerId}]:</b> adj_off=[<b>{rapmOff.toFixed(2)}</b>], adj_def=[<b>{rapmDef.toFixed(2)}</b>]
        <ul>
          <li>RAPM contribution: off=[<b>{rawRapmOff.toFixed(2)}</b>], def=[<b>{rawRapmDef.toFixed(2)}</b>], total=[<b>{totalRawRapm.toFixed(2)}</b>]</li>
          <li>Priors contribution: off=[<b>{priorOff.toFixed(2)}</b>], def=[<b>{priorDef.toFixed(2)}</b>], total=[<b>{totalPrior.toFixed(2)}</b>]</li>
            {priorDiagListEl}
        </ul>
        (<b>More player diagnostics to come...</b>)<br/>(<a href="#" onClick={(event) => { event.preventDefault(); gotoGlobalDiags() }}>Scroll to global RAPM diagnostics</a>)
      </span>;
    }
    catch (err) { //Temp issue during reprocessing
      return <span>Recalculating diags, pending {err.message}</span>;
    }
});

export default RapmPlayerDiagView;
