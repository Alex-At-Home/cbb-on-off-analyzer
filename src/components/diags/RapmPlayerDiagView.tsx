// React imports:
import React, { useState } from 'react';

import _ from "lodash";

// Next imports:
import { NextPage } from 'next';

// Utils
import { RapmInfo, RapmPlayerContext, RapmPreProcDiagnostics, RapmProcessingInputs, RapmUtils } from "../../utils/stats/RapmUtils";

type Props = {
  rapmInfo: RapmInfo,
  player: Record<string, any>,
  globalRef: React.RefObject<HTMLDivElement>
};

const RapmPlayerDiagView: React.FunctionComponent<Props> = (({rapmInfo, player, globalRef}) => {
  try {
    const ctx = rapmInfo.ctx;
    const offWeights = rapmInfo.offWeights.valueOf();
    const offInputs = rapmInfo.offInputs;
    const defInputs = rapmInfo.defInputs;

    const gotoGlobalDiags = () => {
      if (globalRef.current) {
        globalRef.current.scrollIntoView({
          behavior: 'smooth',
          block: 'start',
        });
      }
    };

    const rapmOff = (player.rapm?.off_adj_ppp?.value || 0);
    const rapmDef = (player.rapm?.def_adj_ppp?.value || 0);

    const col = ctx.playerToCol[player.playerId];

    const totalOffPoss = ctx.teamInfo?.off_poss?.value;
    const teamOffAdj = ((ctx.teamInfo.off_adj_ppp?.value || ctx.avgEfficiency) - ctx.avgEfficiency);
    const teamDefAdj = ((ctx.teamInfo.def_adj_ppp?.value || ctx.avgEfficiency) - ctx.avgEfficiency);

    const offPoss = offInputs.playerPossPcts[col]!;
    const defPoss = defInputs.playerPossPcts[col]!;
    const offPossPctStr = (100.0*(offInputs.playerPossPcts[col]!)).toFixed(0);

    // Prior (luck adjusted o/drtg)
    const offPrior = ctx.priorInfo.playersWeak?.[col]?.off_adj_ppp || 0;
    const defPrior = ctx.priorInfo.playersWeak?.[col]?.def_adj_ppp || 0;

    const offUnbiasRapm = offInputs.rapmRawAdjPpp[col];
    const defUnbiasRapm = defInputs.rapmRawAdjPpp[col];

    const buildPrior = (input: RapmProcessingInputs) => {
      const vec = input.rapmRawAdjPpp;
      return _.reduce(vec, (acc, n: number, i: number) => acc + n*input.playerPossPcts[i]!) || 0
    };
    const [ sigmaRapmOff, sigmaRapmDef ]  = [ buildPrior(offInputs), buildPrior(defInputs) ];
    const offPriorTotalDiff = teamOffAdj - sigmaRapmOff;
    const defPriorTotalDiff = teamDefAdj - sigmaRapmDef;
    const offPriorContrib = rapmOff - offUnbiasRapm;
    const defPriorContrib = rapmDef - defUnbiasRapm;

    const maybeAdaptiveWeight = rapmInfo?.preProcDiags?.adaptiveCorrelWeights?.[col] || 0.0;

    const detailedInfoPost = <ul>
      <li>We calculate a team adjustment (off=[<b>{offPriorTotalDiff.toFixed(2)}</b>] def=[<b>{defPriorTotalDiff.toFixed(2)}</b>]) to reduce/remove the
      the delta between total adjusted efficiency and RAPM (due to the regression factor): eg compare <em>observed</em> (off=[<b>{teamOffAdj.toFixed(2)}</b>] def=[<b>{teamDefAdj.toFixed(2)}</b>])
      vs <em>derived solely from RAPM</em> (off=[<b>{sigmaRapmOff.toFixed(2)}</b>] def=[<b>{sigmaRapmDef.toFixed(2)}</b>]).
      </li>
      <li>Then we calculate a player's contribution to this team total - currently this is a fraction of
      "Adj Rtg+": off=[<b>{offPrior.toFixed(2)}</b>] def=[<b>{defPrior.toFixed(2)}</b>]), ...
      </li>
      <li>
      ... chosen so that a minutes-weighted average of the ratings sums to the team value: off=[<b>{offPriorContrib.toFixed(2)}</b>], def=[<b>{defPriorContrib.toFixed(2)}</b>]
        <ul>
          <li>
          <em> (eg incorporating the % on floor [<b>{offPossPctStr}%</b>] (of [<b>{totalOffPoss}</b>] poss,
             this is an off=[<b>{(offPriorContrib*offPoss).toFixed(2)}</b>] def=[<b>{(defPriorContrib*defPoss).toFixed(2)}</b>]
            "slice" of the team total of off=[<b>{offPriorTotalDiff.toFixed(2)}</b>] def=[<b>{defPriorTotalDiff.toFixed(2)}</b>])</em>
          </li>
        </ul>
      </li>
    </ul>;

    const adaptiveWeight = ctx.priorInfo.strongWeight >= 0 ? ctx.priorInfo.strongWeight : maybeAdaptiveWeight;
    const rapmPriorOverrideInfo = ctx.priorInfo.strongWeight >= 0 ?
      <span> (hand-overwritten to <b>{(adaptiveWeight).toFixed(2)}</b>)</span> : null;

    const detailedInfoPre = <ul>
      <li>To combat the tendency of RAPM to over-share the contribution of the strongest players amongst their typical team-mates,
      we take the weighted average player correlation (see "Player correlation table" in the Global Diagnostics below) [<b>{(maybeAdaptiveWeight).toFixed(2)}</b>]{rapmPriorOverrideInfo}, ...
      </li>
      <li>... and use that % of the player's "Adj Rating+" ([<b>{(adaptiveWeight).toFixed(2)}</b>]*[<b>{(offPrior).toFixed(2)}</b>]) = [<b>{(adaptiveWeight*offPrior).toFixed(2)}</b>] as a prior in the RAPM calculation.
      </li>
      <ul>
        <li><i>(Currently we only do this for offense because ORtg is a much more reliable individual stat than DRtg)</i></li>
      </ul>
      <li>ie Off RAPM [<b>{offUnbiasRapm.toFixed(2)}</b>] = Raw RAPM [<b>{(offUnbiasRapm - adaptiveWeight*offPrior).toFixed(2)}</b>] + Prior [<b>{(adaptiveWeight*offPrior).toFixed(2)}</b>]</li>
    </ul>;

    const totalPrior = offPriorContrib - defPriorContrib;
    const totalRawRapm = offUnbiasRapm - defUnbiasRapm;

    //TODO: play around with player RAPM 
    const offMatrixInputs = rapmInfo.offInputs.solnMatrix.valueOf()[col];
    const [ offAdjPoss, defAdjPoss ] = RapmUtils.calcLineupOutputs(
      "adj_ppp", ctx.avgEfficiency, ctx.avgEfficiency, ctx, ctx.priorInfo.adaptiveCorrelWeights, false
    );
    const rapmDiag = offMatrixInputs.map((p, n) => { return {
      lineupId: n,
      lineupVal: offAdjPoss[n]!,
      matrixVal: p,
      contrib: p*offAdjPoss[n]!
    }; });
    const goodOnLineups = _.chain(rapmDiag).filter(d => d.matrixVal > 0 && d.contrib > 0).sortBy([d => -1*d.contrib]).value();
    const badOnLineups = _.chain(rapmDiag).filter(d => d.matrixVal > 0 && d.contrib < 0).sortBy([d => d.contrib]).value();
    const goodOffLineups = _.chain(rapmDiag).filter(d => d.matrixVal < 0 && d.contrib > 0).sortBy([d => -1*d.contrib]).value();
    const badOffLineups = _.chain(rapmDiag).filter(d => d.matrixVal < 0 && d.contrib < 0).sortBy([d => d.contrib]).value();
    const goodOnTotal = _.chain(goodOnLineups).sumBy(d => d.contrib).value().toFixed(2);
    const goodOnOther = _.chain(goodOnLineups).drop(5).sumBy(d => d.contrib).value().toFixed(2);
    const badOnTotal = _.chain(badOnLineups).sumBy(d => d.contrib).value().toFixed(2);
    const badOnOther = _.chain(badOnLineups).drop(5).sumBy(d => d.contrib).value().toFixed(2);
    const goodOffTotal = _.chain(goodOffLineups).sumBy(d => d.contrib).value().toFixed(2);
    const goodOffOther = _.chain(goodOffLineups).drop(5).sumBy(d => d.contrib).value().toFixed(2);
    const badOffTotal = _.chain(badOffLineups).sumBy(d => d.contrib).value().toFixed(2);
    const badOffOther = _.chain(badOffLineups).drop(5).sumBy(d => d.contrib).value().toFixed(2);
    const test = <span>
      Good On: {_.take(goodOnLineups, 5).map(d => d.contrib.toFixed(2)).join(" ; ")} Other=[{goodOnOther}, {goodOnLineups.length - 5}] Total=[{goodOnTotal}]<br/>
      Bad On: {_.take(badOnLineups, 5).map(d => d.contrib.toFixed(2)).join(" ; ")} Other=[{badOnOther}, {badOnLineups.length - 5}] Total=[{badOnTotal}]<br/>
      Good Off: {_.take(goodOffLineups, 5).map(d => d.contrib.toFixed(2)).join(" ; ")} Other=[{goodOffOther}, {goodOffLineups.length - 5}] Total=[{goodOffTotal}]<br/>
      Bad Off: {_.take(badOffLineups, 5).map(d => d.contrib.toFixed(2)).join(" ; ")} Other=[{badOffOther}, {badOffLineups.length - 5}] Total=[{badOffTotal}]<br/>
    </span>;

    return <span>
        <b>RAPM diagnostics for [{player.playerId}]:</b> adj_off=[<b>{rapmOff.toFixed(2)}</b>], adj_def=[<b>{rapmDef.toFixed(2)}</b>] =
        <ul>
          <li>RAPM contribution: off=[<b>{offUnbiasRapm.toFixed(2)}</b>], def=[<b>{defUnbiasRapm.toFixed(2)}</b>], total=[<b>{totalRawRapm.toFixed(2)}</b>]</li>
            {detailedInfoPre}
          <li>&nbsp;+ POST RAPM adjustment: off=[<b>{offPriorContrib.toFixed(2)}</b>], def=[<b>{defPriorContrib.toFixed(2)}</b>], total=[<b>{totalPrior.toFixed(2)}</b>]</li>
            {detailedInfoPost}
        </ul>
        {test}<br/>
        (<b>More player diagnostics to come...</b>)<br/>(<a href="#" onClick={(event) => { event.preventDefault(); gotoGlobalDiags() }}>Scroll to global RAPM diagnostics</a>)
      </span>;
    }
    catch (err) { //Temp issue during reprocessing
      return <span>Recalculating diags, pending {err.message}</span>;
    }
});

export default RapmPlayerDiagView;
