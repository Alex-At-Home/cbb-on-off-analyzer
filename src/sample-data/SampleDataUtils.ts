import _ from "lodash";

export class SampleDataUtils {

  /** For snapshot testing, provides a more useful output than a giant list of stats */
  static summarizeEnrichedApiResponse(exampleRawResponse: Record<string, any>, mutableSample: boolean = false) {
    const expectedFields1 = mutableSample ? null : new Set(_.keys(exampleRawResponse));

    return {
      test: (val: any) => val && (val.off_efg ||
          val.hasOwnProperty("dRtg") || val.hasOwnProperty("oRtg") ||
          val.hasOwnProperty("deltaOffPpp") || val.hasOwnProperty("deltaDefPpp")
        ),
        //(pick a field that all stat sets have - plus special case for: diagnostics)
      print: (val: any, serialize: ((val: any) => string), indent: ((val: any) => string)) => {
        const expectedFields = expectedFields1 || new Set(_.keys(exampleRawResponse));
        if (val.off_efg) {
          return "{\n" +
            _.chain(val).toPairs().filter(kv => {
              return !expectedFields.has(kv[0]) //derived field)
                || (!kv[1]?.hasOwnProperty("value") && exampleRawResponse[kv[0]]?.hasOwnProperty("value")) //missing value
                || (kv[1]?.hasOwnProperty("value") && _.keys(kv[1]).length > 1) //(has extra properties)
            }).sortBy(0).map(kv => {
              return indent(`${kv[0]}: ` + serialize(kv[1]))
            }).value().join("\n\n") +
          "}";
        } else if (val.hasOwnProperty("dRtg") || val.hasOwnProperty("oRtg")) {
          return JSON.stringify(_.pick(val, [
            "oRtg", "adjORtg", "adjORtgPlus", "dRtg", "adjDRtg", "adjDRtgPlus"
          ]), null , 3);
        } else if (val.hasOwnProperty("deltaOffPpp") || val.hasOwnProperty("deltaDefPpp")) {
          return JSON.stringify(_.pick(val, [
            "deltaOffPpp", "deltaOffEfg", "deltaOffAdjEff", "deltaDefPpp", "deltaDefEfg", "deltaDefAdjEff"
          ]), null , 3);
        } else {
          return "???";
        }
      }
    }
  }


  /** From easier to read templates builds a bucket list */
  static buildResponseFromTemplateLineup(input: Record<string, any>) {
    const lineupAsObj = SampleDataUtils.buildResponseFromTemplateTeam(input);
    return _.toPairs(lineupAsObj).map(kv => {
      return { key: kv[0], ...(kv[1] as Record<string, any>) };
    })
  }

  /** From easier to read templates builds a bucket list */
  static buildResponseFromTemplateTeam(input: Record<string, any>) {
    const teamAsObj = _.transform(_.toPairs(input), (acc, inElKv) => {
      _.forEach(_.toPairs(inElKv[1] as Record<string, any>), (kv) => {
        _.forEach(_.toPairs(kv[1] as Record<string, any>), (deepKv) => {
          if (inElKv[0] != "main") {
            _.set(acc, `${kv[0]}.${deepKv[0]}${inElKv[0]}.value`, deepKv[1]);
          } else {
            _.set(acc, `${deepKv[0]}.${kv[0]}`, deepKv[1]);
          }
        });
      });
    }, {} as Record<string, any>);
    return teamAsObj;
  }

  /** From easier to read templates builds a bucket list */
  static buildResponseFromTemplatePlayer(input: Record<string, any>) {
    const flatPlayersAsObj = SampleDataUtils.buildResponseFromTemplateTeam(input);
    return _.transform(["baseline", "on", "off"], (acc, v) => {
      acc[v] = { player: {
        buckets: _.chain(flatPlayersAsObj).toPairs().filter(kv => {
          return kv[0].indexOf(v) == 0;
        }).map(kv => {
          const key = kv[0].substring(v.length + 2); //(+2 for ": ")
          return { ...(kv[1]), key: key }
        }).value()
      } };
    }, {} as Record<string, any>);
  }

  /** Builds an easier to manipulate template for the different responses, from an actual response */
  static buildTemplateFromResponseTeam(input: Record<string, Record<string, any>>) {
    return SampleDataUtils.buildTemplateFromResponseLineup(
      _.chain(input).toPairs().map((kv: [string, any]) => {
        return { key: kv[0], ...(kv[1]) }
      }).value()
    );
  }

  /** Builds an easier to manipulate template for the different responses, from an actual response */
  static buildTemplateFromResponsePlayer(input: Record<string, Record<string, any>>) {
    return SampleDataUtils.buildTemplateFromResponseLineup(
      _.chain(input).toPairs().flatMap((kv: [string, any]) => {
        const players = (kv[1]?.player?.buckets || []) as Array<Record<string, any>>;
        return players.map(p => { return { ...p, key: `${kv[0]}: ${p.key}` } });
      }).value()
    );
  }

  /** Builds an easier to manipulate template for the different responses, from an actual response */
  static buildTemplateFromResponseLineup(input: Array<Record<string, any>>, filter?: Set<String>) {
    const preSorted = _.transform(input, (acc, inEl) => {
      const recordId = inEl?.key || "unknown";
      const injectInfo = (k: string, prefix: string) => {
        const topLevelKey = k.substring(prefix.length);
        const realPrefix = prefix || "main";
        if (prefix) {
          const existingObj = _.get(acc, `${topLevelKey}.${recordId}`) || {};
          existingObj[realPrefix] = inEl[k]?.value;
          const sortedObj = _(existingObj).toPairs().sortBy(0).reverse().fromPairs().value()
          _.set(acc, `${topLevelKey}.${recordId}`, sortedObj);
        } else {
          _.set(acc, `main.${topLevelKey}.${recordId}`, inEl[k]);
        }
      };

      if (!filter || filter.has(recordId)) {
        _.keys(inEl).filter(k => k != "key").forEach(k => {
          const stat = _.find([
            "oppo_total_off_", "oppo_total_def_",
            "oppo_off_", "oppo_def_",

            "team_total_off_", "team_total_def_",
            "team_off_", "team_def_",

            "total_off_", "total_def_",
            "off_", "def_"
          ], prefix => k.indexOf(prefix) == 0);
          if (stat) {
            injectInfo(k, stat);
          } else {
            injectInfo(k, "");
          }
        });
      }
    }, {} as Record<string, any>);

    // Sort the keys into some sort of logical order
    return _.chain(preSorted).toPairs().sortBy([
      (kv: [ string , any ]) => {
        const key = kv[0];
        if (key == "main") {
          return 100;
        } else if ((key.indexOf("assist") >= 0) || (key.indexOf("to") >= 0) || (key.indexOf("ast") >= 0)) {
          return 4;
        } else if ((key.indexOf("ft") == 0) || (key.indexOf("fg") == 0) || (key == "efg")) {
          return 1;
        } else if ((key.indexOf("2p") == 0) || (key.indexOf("3p") == 0)) {
          return 2;
        } else if ((key.indexOf("rb") > 0)) {
          return 3;
        } else if ((key.indexOf("adj") >= 0)) {
          return 5;
        } else { //misc other stuff
          return 6;
        }
      },
      (kv: [ string , any ]) => {
        return kv[0];
      },
    ]).fromPairs().value();
  }
}
