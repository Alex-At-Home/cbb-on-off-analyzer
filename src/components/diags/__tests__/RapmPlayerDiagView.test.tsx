import renderer from 'react-test-renderer';
import React from 'react';
import RapmPlayerDiagView from '../RapmPlayerDiagView';

// Needed to build the data
import { sampleLineupStatsResponse } from "../../../sample-data/sampleLineupStatsResponse";
import { LineupUtils } from "../../../utils/stats/LineupUtils";
import { RapmUtils } from "../../../utils/stats/RapmUtils";
import { semiRealRapmResults } from "../../../utils/stats/__tests__/RapmUtils.test";
import { StatModels, LineupStatSet } from '../../../utils/StatModels';

describe("RapmPlayerDiagView", () => {

  const lineupReport = {
    lineups: sampleLineupStatsResponse.responses[0].aggregations.lineups.buckets as LineupStatSet[],
    avgEff: 100.0,
    error_code: "test"
  };
  const globalRapmDiagRef = React.createRef<HTMLDivElement>();

  test("RapmPlayerDiagView - should create snapshot", () => {
    const [ offResults, defResults ] = RapmUtils.pickRidgeRegression(
      semiRealRapmResults.testOffWeights, semiRealRapmResults.testDefWeights, semiRealRapmResults.testContext, undefined, true
    );
    const onOffReport = LineupUtils.lineupToTeamReport(lineupReport);

    RapmUtils.injectRapmIntoPlayers(
      onOffReport.players || [], offResults, defResults, {}, semiRealRapmResults.testContext, undefined
    );
    const rapmInfo = {
      ctx: semiRealRapmResults.testContext,
      preProcDiags: RapmUtils.calcCollinearityDiag(
        semiRealRapmResults.testOffWeights, semiRealRapmResults.testContext
      ),
      offWeights: semiRealRapmResults.testOffWeights,
      defWeights: semiRealRapmResults.testDefWeights,
      offInputs: offResults,
      defInputs: defResults,
      enrichedPlayers: []
    };

    const component = renderer.create(<RapmPlayerDiagView
      globalRef={globalRapmDiagRef}
      rapmInfo={rapmInfo}
      player={onOffReport.players![0]!}
    />);

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
