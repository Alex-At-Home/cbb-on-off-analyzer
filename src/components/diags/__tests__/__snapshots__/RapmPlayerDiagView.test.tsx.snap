// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RapmPlayerDiagView RapmPlayerDiagView - should create snapshot 1`] = `
<span>
  <b>
    RAPM diagnostics for [
    Wiggins, Aaron
    ]:
  </b>
   adj_off=[
  <b>
    5.15
  </b>
  ], adj_def=[
  <b>
    -4.94
  </b>
  ] =
  <ul>
    <li>
      RAPM contribution: off=[
      <b>
        5.15
      </b>
      ], def=[
      <b>
        -4.48
      </b>
      ], total=[
      <b>
        9.62
      </b>
      ]
    </li>
    <ul>
      <li>
        To combat the tendency of RAPM to over-share the contribution of the strongest players amongst their typical team-mates, we take the weighted average player correlation (see "Player correlation table" in the Global Diagnostics below) [
        <b>
          1.09
        </b>
        ]
        <span>
           (hand-overwritten to 
          <b>
            0.50
          </b>
          )
        </span>
        , ...
      </li>
      <li>
        ... and use that % of the player's "Adj Rating+" ([
        <b>
          0.50
        </b>
        ]*[
        <b>
          4.00
        </b>
        ]) = [
        <b>
          2.00
        </b>
        ] as a prior in the RAPM calculation.
      </li>
      <ul>
        <li>
          <i>
            (Currently we only do this for offense because ORtg is a much more reliable individual stat than DRtg)
          </i>
        </li>
      </ul>
      <li>
        ie Off RAPM [
        <b>
          5.15
        </b>
        ] = Raw RAPM [
        <b>
          3.15
        </b>
        ] + Prior [
        <b>
          2.00
        </b>
        ]
      </li>
    </ul>
    <span>
      <li>
        Â + POST RAPM adjustment: off=[
        <b>
          0.00
        </b>
        ], def=[
        <b>
          -0.47
        </b>
        ], total=[
        <b>
          0.47
        </b>
        ]
      </li>
      <ul>
        <li>
          We calculate a team adjustment (off=[
          <b>
            -10.15
          </b>
          ] def=[
          <b>
            -2.07
          </b>
          ]) to reduce/remove the the delta between total adjusted efficiency and RAPM (due to the regression factor): eg compare 
          <em>
            observed
          </em>
           (off=[
          <b>
            10.00
          </b>
          ] def=[
          <b>
            -20.00
          </b>
          ]) vs 
          <em>
            derived solely from RAPM
          </em>
           (off=[
          <b>
            20.15
          </b>
          ] def=[
          <b>
            -17.93
          </b>
          ]).
        </li>
        <span>
          <li>
            Then we calculate a player's contribution to this team total - currently this is a fraction of "Adj Rtg+": off=[
            <b>
              4.00
            </b>
            ] def=[
            <b>
              -4.00
            </b>
            ]), ...
          </li>
          <li>
            ... chosen so that a minutes-weighted average of the ratings sums to the team value: off=[
            <b>
              0.00
            </b>
            ], def=[
            <b>
              -0.47
            </b>
            ]
            <ul>
              <li>
                <em>
                   (eg incorporating the % on floor [
                  <b>
                    82
                    %
                  </b>
                  ] (of [
                  <b>
                    101
                  </b>
                  ] poss, this is an off=[
                  <b>
                    0.00
                  </b>
                  ] def=[
                  <b>
                    -0.38
                  </b>
                  ] "slice" of the team total of off=[
                  <b>
                    -10.15
                  </b>
                  ] def=[
                  <b>
                    -2.07
                  </b>
                  ])
                </em>
              </li>
            </ul>
          </li>
        </span>
      </ul>
    </span>
  </ul>
  (
  <b>
    More player diagnostics to come...
  </b>
  )
  <br />
  (
  <a
    href="#"
    onClick={[Function]}
  >
    Scroll to global RAPM diagnostics
  </a>
  )
</span>
`;

exports[`RapmUtils RapmUtils - buildPlayerContext 1`] = `
Object {
  "avgEfficiency": 100,
  "colToPlayer": Array [
    "Wiggins, Aaron",
    "Cowan, Anthony",
    "Smith, Jalen",
    "Morsell, Darryl",
    "Ayala, Eric",
    "Scott, Donta",
    "Player, Other",
    "Data, Dummy",
  ],
  "defLineupPoss": 944,
  "numLineups": 5,
  "numPlayers": 8,
  "offLineupPoss": 959,
  "playerToCol": Object {
    "Ayala, Eric": 4,
    "Cowan, Anthony": 1,
    "Data, Dummy": 7,
    "Morsell, Darryl": 3,
    "Player, Other": 6,
    "Scott, Donta": 5,
    "Smith, Jalen": 2,
    "Wiggins, Aaron": 0,
  },
  "priorInfo": Object {
    "includeStrong": Object {},
    "noWeakPrior": false,
    "playersStrong": Array [
      Object {
        "off_adj_ppp": 4.5,
      },
      Object {
        "off_adj_ppp": 5,
      },
      Object {
        "off_adj_ppp": 0,
      },
      Object {
        "off_adj_ppp": 0,
      },
      Object {
        "off_adj_ppp": 0,
      },
      Object {
        "off_adj_ppp": 0,
      },
      Object {
        "off_adj_ppp": 0,
      },
      Object {
        "off_adj_ppp": 0,
      },
    ],
    "playersWeak": Array [
      Object {
        "def_adj_ppp": -4.5,
        "off_adj_ppp": 4.5,
      },
      Object {
        "def_adj_ppp": -5,
        "off_adj_ppp": 5,
      },
      Object {
        "def_adj_ppp": 0,
        "off_adj_ppp": 0,
      },
      Object {
        "def_adj_ppp": 0,
        "off_adj_ppp": 0,
      },
      Object {
        "def_adj_ppp": 0,
        "off_adj_ppp": 0,
      },
      Object {
        "def_adj_ppp": 0,
        "off_adj_ppp": 0,
      },
      Object {
        "def_adj_ppp": 0,
        "off_adj_ppp": 0,
      },
      Object {
        "def_adj_ppp": 0,
        "off_adj_ppp": 0,
      },
    ],
    "strongWeight": -1,
    "useRecursiveWeakPrior": false,
  },
  "removalPct": 0,
  "removedPlayers": Object {},
  "unbiasWeight": 0,
}
`;

exports[`RapmUtils RapmUtils - buildPlayerContext 2`] = `
Object {
  "avgEfficiency": 100,
  "colToPlayer": Array [
    "Wiggins, Aaron",
    "Cowan, Anthony",
    "Smith, Jalen",
    "Morsell, Darryl",
    "Ayala, Eric",
    "Scott, Donta",
  ],
  "defLineupPoss": 794,
  "numLineups": 3,
  "numPlayers": 6,
  "offLineupPoss": 809,
  "playerToCol": Object {
    "Ayala, Eric": 4,
    "Cowan, Anthony": 1,
    "Morsell, Darryl": 3,
    "Scott, Donta": 5,
    "Smith, Jalen": 2,
    "Wiggins, Aaron": 0,
  },
  "priorInfo": Object {
    "includeStrong": Object {},
    "noWeakPrior": false,
    "playersStrong": Array [
      Object {
        "off_adj_ppp": 4.5,
      },
      Object {
        "off_adj_ppp": 5,
      },
      Object {
        "off_adj_ppp": 0,
      },
      Object {
        "off_adj_ppp": 0,
      },
      Object {
        "off_adj_ppp": 0,
      },
      Object {
        "off_adj_ppp": 0,
      },
    ],
    "playersWeak": Array [
      Object {
        "def_adj_ppp": -4.5,
        "off_adj_ppp": 4.5,
      },
      Object {
        "def_adj_ppp": -5,
        "off_adj_ppp": 5,
      },
      Object {
        "def_adj_ppp": 0,
        "off_adj_ppp": 0,
      },
      Object {
        "def_adj_ppp": 0,
        "off_adj_ppp": 0,
      },
      Object {
        "def_adj_ppp": 0,
        "off_adj_ppp": 0,
      },
      Object {
        "def_adj_ppp": 0,
        "off_adj_ppp": 0,
      },
    ],
    "strongWeight": -1,
    "useRecursiveWeakPrior": false,
  },
  "removalPct": 0.2,
  "removedPlayers": Object {
    "Data, Dummy": Array [
      0.05254860746190226,
      0.05254860746190226,
    ],
    "Player, Other": Array [
      0.1576458223857068,
      0.1576458223857068,
    ],
  },
  "unbiasWeight": 0,
}
`;
